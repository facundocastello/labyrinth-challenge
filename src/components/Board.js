import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import { maxSize } from '../utils/constants';
import Backdrop from './Backdrop';
import GameTypography from './GameTypography';
import Slot from './Slot';

const Horizontal = styled.div`
  display: flex;
`;

const Vertical = styled.div`
  display: flex;
  flex-direction: column;
`;

const BoardContainer = styled.div`
  border: 1px solid black;
  width: fit-content;
`;

export default function Board( {
  editting, blockedPositions, endPosition, startPosition,
  size, slotClicked, handleEndGame, win, lost, time
} ) {
  const [boardItems, setBoardItems] = useState([]);

  useEffect(() => {
    var newBoardItems = new Array( size );
    for ( let x = 0; x < size; x++ ) {
      newBoardItems[x] = new Array( size );
      for ( let y = 0; y < size; y++ ) {
        if (( x * size + y ) === endPosition ) {
          newBoardItems[x][y] = 'end';
          continue;
        }
        if (( x * size + y ) === startPosition ) {
          newBoardItems[x][y] = 'start';
          continue;
        }
        const isBlocked = blockedPositions.find(( blockedPosition ) =>
          ( x * size + y ) === blockedPosition ) > -1;

        if ( isBlocked ) {
          newBoardItems[x][y] = 'blocked';
          continue;
        }
        newBoardItems[x][y] = 'empty';
      }
    }
    setBoardItems( newBoardItems );

  }, [blockedPositions, endPosition, size, startPosition]);

  const truncate = ( str, length ) =>
    str.length > length ? str.substring( 0, length ) + '...' : str;

  return <BoardContainer data-testid="board">
    {!editting && ( win || lost ) &&
      <Backdrop size={size}>
        <GameTypography win={win} lost={lost}>
          {win ? ( size === maxSize ? 'Game Over' : 'You Win' ) : lost ? 'You Lose' : ''}
        </GameTypography>
        {win && <GameTypography win={win} lost={lost}>
          Your Time: {truncate(( time ) + '' )}s
        </GameTypography>}
        <GameTypography data-testid="end-game" onClick={handleEndGame}>
          {win ? ( size === maxSize ? '' : 'Next level' ) : lost ? 'Play again' : ''}
        </GameTypography>
      </Backdrop>
    }
    <Vertical>
      {boardItems.map(( row, index ) => <Horizontal key={`row-${index}`}>
        {row && row.map(( slot, slotIndex ) => <Slot
          data-testid={`slot-${index * size + slotIndex}`}
          onClick={() => slotClicked( size * index + slotIndex )}
          type={slot} key={`slot-${index}-${slotIndex}`} /> )}
      </Horizontal>
      )}
    </Vertical>
  </BoardContainer>;
}
