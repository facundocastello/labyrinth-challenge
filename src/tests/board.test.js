import React from 'react';
import Board from '../components/Board';
import { create } from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';

const renderBoard = ( size, blockedPositions, startPosition, endPosition ) => {
  const { getByTestId } = render( <Board
    startPosition={startPosition} endPosition={endPosition} blockedPositions={blockedPositions}
    size={size} /> );
  for ( let index = 0; index < size * size; index++ ) {
    const slot = getByTestId( `slot-${index}` );
    let type = 'empty';
    if ( index === endPosition ) { type = 'end'; } else
    if ( index === startPosition ) { type = 'start'; } else
    if ( blockedPositions.includes( index )) type = 'blocked';
    expect( slot ).toHaveAttribute( 'type', type );
  }
};

describe( 'Board Block', () => {
  it( 'renders Board', () => {
    const tree = create(
      <Board blockedPositions={[]} size={5} />
    ).toJSON();
    expect( tree ).toMatchSnapshot();
  } );

  it( 'renders 5x5 board with 1 blocked', () => {
    const size = 5;
    const blockedPositions = [1];
    const startPosition = 0;
    const endPosition = 5;
    renderBoard( size, blockedPositions, startPosition, endPosition );
  } );

  it( 'render 5x5 board with 1 blocked and click on first slot', () => {
    const size = 5;
    const blockedPositions = [1];
    const startPosition = 0;
    const endPosition = 5;
    const index = 5;
    const sortClicked = jest.fn();
    const { getByTestId } = render( <Board
      slotClicked={sortClicked} startPosition={startPosition} endPosition={endPosition}
      blockedPositions={blockedPositions} size={size} /> );
    fireEvent.click( getByTestId( `slot-${index}` ));
    expect( sortClicked ).toHaveBeenCalledWith( index );
  } );

  it( 'renders 20x20 board with 10 blocked', () => {
    const size = 20;
    const blockedPositions = [1, 2, 3, 4, 6, 7, 8, 9, 90];
    const startPosition = 0;
    const endPosition = 5;
    renderBoard( size, blockedPositions, startPosition, endPosition );
  } );

  it( 'renders board with the message You Lose', () => {
    const size = 5;
    const blockedPositions = [];
    const startPosition = 0;
    const endPosition = 5;
    const { getByText } = render( <Board
      lost={true}
      startPosition={startPosition} endPosition={endPosition} blockedPositions={blockedPositions}
      size={size} /> );
    expect( getByText( 'You Lose' )).not.toBeNull();
    expect( getByText( 'Play again' )).not.toBeNull();
  } );

  it( 'renders board with the message You Win and the time', () => {
    const size = 5;
    const blockedPositions = [];
    const startPosition = 0;
    const endPosition = 5;
    const time = 5;
    const { getByText } = render( <Board
      win={true} time={time}
      startPosition={startPosition} endPosition={endPosition} blockedPositions={blockedPositions}
      size={size} /> );
    expect( getByText( 'You Win' )).not.toBeNull();
    expect( getByText( `Your Time: ${time}s` )).not.toBeNull();
    expect( getByText( 'Next level' )).not.toBeNull();
  } );
} );