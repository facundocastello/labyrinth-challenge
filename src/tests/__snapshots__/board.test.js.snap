// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board Block renders Board 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 {
  border: 1px solid black;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

<div
  className="c0"
  data-testid="board"
>
  <div
    className="c1"
  />
</div>
`;
